src/terrain/terrain.hpp

```diff
@@ -857,3 +584,15 @@ class Terrain : voxel_utility::VoxelBase{
 };
 
 } // namespace terrain
+
+inline uint32_t terrain::Chunk::get_voxel(int x, int y, int z) const {
+    return ter_->get_voxel(
+        x + Cx_ * Chunk::SIZE, y + Cy_ * Chunk::SIZE, z + Cz_ * Chunk::SIZE
+    );
+}
+
+inline uint16_t terrain::Chunk::get_voxel_color_id(int x, int y, int z) const {
+    return ter_->get_voxel_color_id(
+        x + Cx_ * Chunk::SIZE, y + Cy_ * Chunk::SIZE, z + Cz_ * Chunk::SIZE
+    );
+}
```

Comment:
Why is this not defined inline in the `chunk.hpp` file

Reply:
weve had this conversation multipul times. ter_ is of type terrain, and terrain.hpp cannot be included in the chunk header.

Reply:
I think this reduced mech generation from 2:00 to like 30 seconds

Reply:
I highly doubt that. Especially if you aren't compiling with `-O3` and [LTO enabled](https://cmake.org/cmake/help/latest/module/CheckIPOSupported.html) then that result is worthless.

## Explanation
Both `Chunk::get_voxel` and `Chunk::get_voxel_color_id` eventually do a hashmap lookup. You have a lot of voxels, so likely the elements of the hashmap won't be in your L1 or L2 cache, and probably not even your L3, meaning the lookup will take several orders of magnitude longer than any computation done.

Now, you might say, "But you're doing a jump if you put this in the CPP file". And you would be right. But this is an **unconditional jump**. Meaning that there is no branch prediction. The CPU will look ahead, see "oh we're doing a call", and prefetch this procedure and put it into the L1i cache. This procedure will also likely be hot, so it will probably be in either L2 or L3 at all times. Even if either `Chunk::get_voxel*` or `Terrain::get_voxel*` are not cached, they will be prefetched and so the longest wait here will be the hashmap lookup by several orders of magnitude.

And if you turn on LTO, the whole last paragraph will be moot as `Chunk::get_voxel*` will be inlined at link time.

Reply:
ill, test it, but there is rarely a look up
The function has three static vars that should avoid a hashmap lookup